### PR TITLE
Make paginate more efficient with many pages.

### PR DESCRIPTION
**Pull request by dourvaris**
Current paginate iterates over every single page possible, and discards most of them, for large page sets > 50,000 this can be quite slow and noticeable.

This patch should make paginating on many pages much faster since it won't iterate over every single page possible, it should make it constant time more or less.

Speed for 100,000 pages:

``` ipython
(old)
In [14]: %timeit list(flask_sqlalchemy.Pagination(None, 4, 10, 1000000, []).iter_pages())
10 loops, best of 3: 81.6 ms per loop

(new)
In [11]: %timeit list(flask_sqlalchemy.Pagination(None, 4, 10, 1000000, []).iter_pages())
100000 loops, best of 3: 11 µs per loop

```

Also added a bunch of tests for edge cases.

### CODE REVIEW

File: flask_sqlalchemy/__init__.py [hunk 1]
```diff
@@ -390,16 +390,28 @@ def iter_pages(self, left_edge=2, left_current=2,
               </div>
             {% endmacro %}
         """
+
+        pages = []
+        pages += range(
```
Comment (justanr) [inline flask_sqlalchemy/__init__.py]: This breaks for Python 3. Why not `pages.extend` or  just `itertools.chain(...)` since there's not a reason to keep them all in memory anyways.

**Comment by dourvaris** [comment]
Is there any interest in this pr at all?

**Comment by dhhagan** [comment]
Has/is this going to be merged..?

**Comment by davidism** [comment]
fixed in #1087
